File: Components/Templating/BaseTokenReplace.vb
'
' DotNetNuke® - http://www.dotnetnuke.com
' Copyright (c) 2002-2010 by DotNetNuke Corp. 
'
' Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
' documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
' the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
' to permit persons to whom the Software is furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all copies or substantial portions 
' of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
' THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
' CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
' DEALINGS IN THE SOFTWARE.
'

Imports DotNetNuke
Imports DotNetNuke.Entities.Modules
Imports DotNetNuke.Entities.Portals
Imports DotNetNuke.Entities.Tabs
Imports DotNetNuke.Entities.Users
Imports DotNetNuke.Entities.Profile
Imports DotNetNuke.Services.Localization
Imports System.Globalization
Imports System.Web
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Reflection
Imports DotNetNuke.UI.Utilities


Namespace Templating

 ''' <summary>
 ''' The BaseTokenReplace class provides the tokenization of tokens formatted  
 ''' [object:property] or [object:property|format|ifEmpty] or [custom:no] within a string
 ''' with the appropriate current property/custom values.
 ''' </summary>
 ''' <remarks></remarks>
 Public MustInherit Class BaseTokenReplace

#Region "Regexp Tokenizer"


  'Private Const ExpressionDefault As String = "(?:\[(?:(?<object>[^\]\[:]+):(?<property>[^\]\[\|]+))(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])|(?<text>\[[^\]\[]+\])|(?<text>[^\]\[]+)"
  Private Const ExpressionDefault As String = "\[(?<object>[^\]\[:\r\n]+):(?<property>[^\]\[\|\r\n]+)((\])|(\|(?<format>[^\]\[\r\n]+)((\])|(\|(?<ifEmpty>[^\]\[\r\n]+)))))"

  Private Const ExpressionObjectLess As String = "(?:\[(?:(?<object>[^\]\[:]+):(?<property>[^\]\[\|]+))(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])" _
              & "|(?:(?<object>\[)(?<property>[A-Z0-9._]+)(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])" _
              & "|(?<text>\[[^\]\[]+\])" _
              & "|(?<text>[^\]\[]+)"

  Private _UseObjectLessExpression As Boolean = False

  Protected Property UseObjectLessExpression() As Boolean
   Get
    Return _UseObjectLessExpression
   End Get
   Set(ByVal value As Boolean)
    _UseObjectLessExpression = value
   End Set
  End Property

  Private ReadOnly Property TokenReplaceCacheKey() As String
   Get
    If UseObjectLessExpression Then
     Return "Blog_TokenReplaceRegEx_Objectless"
    Else
     Return "Blog_TokenReplaceRegEx_Default"
    End If
   End Get
  End Property

  Private ReadOnly Property RegExpression() As String
   Get
    If UseObjectLessExpression Then
     Return ExpressionObjectLess
    Else
     Return ExpressionDefault
    End If
   End Get
  End Property

  Protected Const ObjectLessToken As String = "no_object"

  ''' <summary>
  ''' Gets the Regular expression for the token to be replaced
  ''' </summary>
  ''' <value>A regular Expression</value>   
  Protected ReadOnly Property TokenizerRegex() As Regex
   Get
    Dim tokenizer As Regex = CType(DataCache.GetCache(TokenReplaceCacheKey), Regex)
    If tokenizer Is Nothing Then
     tokenizer = New Regex(RegExpression, RegexOptions.Compiled)
     DataCache.SetCache(TokenReplaceCacheKey, tokenizer)
    End If
    Return tokenizer
   End Get
  End Property
#End Region

#Region "Fields & Properties"
  Private _Language As String
  Private _FormatProvider As CultureInfo

  ''' <summary>
  ''' Gets/sets the language to be used, e.g. for date format
  ''' </summary>
  ''' <value>A string, representing the locale</value>
  Public Property Language() As String
   Get
    Return _Language
   End Get
   Set(ByVal value As String)
    _Language = value
    _FormatProvider = New CultureInfo(_Language)
   End Set
  End Property

  ''' <summary>
  ''' Gets the Format provider as Culture info from stored language or current culture
  ''' </summary>
  ''' <value>An CultureInfo</value>
  Protected ReadOnly Property FormatProvider() As System.Globalization.CultureInfo
   Get
    If _FormatProvider Is Nothing Then
     _FormatProvider = System.Threading.Thread.CurrentThread.CurrentUICulture
    End If
    Return _FormatProvider
   End Get
  End Property
#End Region

  Protected Overridable Function ReplaceTokens(ByVal strSourceText As String) As String
   If strSourceText Is Nothing Then Return String.Empty
   Dim Result As New System.Text.StringBuilder
   Return TokenizerRegex.Replace(strSourceText, AddressOf ReplaceTokenMatch)
  End Function

  Private Function ReplaceTokenMatch(m As Match) As String

   Dim strObjectName As String = m.Result("${object}")
   If strObjectName.Length > 0 Then
    If strObjectName = "[" Then strObjectName = ObjectLessToken
    Dim strPropertyName As String = m.Result("${property}")
    Dim strFormat As String = m.Result("${format}")
    Dim strIfEmptyReplacment As String = m.Result("${ifEmpty}")
    Dim strConversion As String = replacedTokenValue(strObjectName, strPropertyName, strFormat)
    If strIfEmptyReplacment.Length > 0 AndAlso strConversion.Length = 0 Then strConversion = strIfEmptyReplacment
    Return strConversion
   Else
    Return m.Value
   End If

  End Function


  Protected MustOverride Function replacedTokenValue(ByVal strObjectName As String, ByVal strPropertyName As String, ByVal strFormat As String) As String

 End Class
End Namespace

File: Server/Core/Templating/BaseTokenReplace.cs

// 
// DotNetNuke® - http://www.dotnetnuke.com
// Copyright (c) 2002-2010 by DotNetNuke Corp. 
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all copies or substantial portions 
// of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
// DEALINGS IN THE SOFTWARE.
// 

using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using DotNetNuke.UI.Utilities;

namespace DotNetNuke.Modules.Blog.Templating
{

  /// <summary>
 /// The BaseTokenReplace class provides the tokenization of tokens formatted  
 /// [object:property] or [object:property|format|ifEmpty] or [custom:no] within a string
 /// with the appropriate current property/custom values.
 /// </summary>
 /// <remarks></remarks>
  public abstract class BaseTokenReplace
  {

    #region Regexp Tokenizer


    // Private Const ExpressionDefault As String = "(?:\[(?:(?<object>[^\]\[:]+):(?<property>[^\]\[\|]+))(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])|(?<text>\[[^\]\[]+\])|(?<text>[^\]\[]+)"
    private const string ExpressionDefault = @"\[(?<object>[^\]\[:\r\n]+):(?<property>[^\]\[\|\r\n]+)((\])|(\|(?<format>[^\]\[\r\n]+)((\])|(\|(?<ifEmpty>[^\]\[\r\n]+)))))";

    private const string ExpressionObjectLess = @"(?:\[(?:(?<object>[^\]\[:]+):(?<property>[^\]\[\|]+))(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])" + @"|(?:(?<object>\[)(?<property>[A-Z0-9._]+)(?:\|(?:(?<format>[^\]\[]+)\|(?<ifEmpty>[^\]\[]+))|\|(?:(?<format>[^\|\]\[]+)))?\])" + @"|(?<text>\[[^\]\[]+\])" + @"|(?<text>[^\]\[]+)";



    private bool _UseObjectLessExpression = false;

    protected bool UseObjectLessExpression
    {
      get
      {
        return _UseObjectLessExpression;
      }
      set
      {
        _UseObjectLessExpression = value;
      }
    }

    private string TokenReplaceCacheKey
    {
      get
      {
        if (UseObjectLessExpression)
        {
          return "Blog_TokenReplaceRegEx_Objectless";
        }
        else
        {
          return "Blog_TokenReplaceRegEx_Default";
        }
      }
    }

    private string RegExpression
    {
      get
      {
        if (UseObjectLessExpression)
        {
          return ExpressionObjectLess;
        }
        else
        {
          return ExpressionDefault;
        }
      }
    }

    protected const string ObjectLessToken = "no_object";

    /// <summary>
  /// Gets the Regular expression for the token to be replaced
  /// </summary>
  /// <value>A regular Expression</value>
    protected Regex TokenizerRegex
    {
      get
      {
        Regex tokenizer = (Regex)DataCache.GetCache(TokenReplaceCacheKey);
        if (tokenizer is null)
        {
          tokenizer = new Regex(RegExpression, RegexOptions.Compiled);
          DataCache.SetCache(TokenReplaceCacheKey, tokenizer);
        }
        return tokenizer;
      }
    }
    #endregion

    #region Fields & Properties
    private string _Language;
    private CultureInfo _FormatProvider;

    /// <summary>
  /// Gets/sets the language to be used, e.g. for date format
  /// </summary>
  /// <value>A string, representing the locale</value>
    public string Language
    {
      get
      {
        return _Language;
      }
      set
      {
        _Language = value;
        _FormatProvider = new CultureInfo(_Language);
      }
    }

    /// <summary>
  /// Gets the Format provider as Culture info from stored language or current culture
  /// </summary>
  /// <value>An CultureInfo</value>
    protected CultureInfo FormatProvider
    {
      get
      {
        if (_FormatProvider is null)
        {
          _FormatProvider = System.Threading.Thread.CurrentThread.CurrentUICulture;
        }
        return _FormatProvider;
      }
    }
    #endregion

    protected virtual string ReplaceTokens(string strSourceText)
    {
      if (strSourceText is null)
        return string.Empty;
      var Result = new StringBuilder();
      return TokenizerRegex.Replace(strSourceText, ReplaceTokenMatch);
    }

    private string ReplaceTokenMatch(Match m)
    {

      string strObjectName = m.Result("${object}");
      if (strObjectName.Length > 0)
      {
        if (strObjectName == "[")
          strObjectName = ObjectLessToken;
        string strPropertyName = m.Result("${property}");
        string strFormat = m.Result("${format}");
        string strIfEmptyReplacment = m.Result("${ifEmpty}");
        string strConversion = replacedTokenValue(strObjectName, strPropertyName, strFormat);
        if (strIfEmptyReplacment.Length > 0 && strConversion.Length == 0)
          strConversion = strIfEmptyReplacment;
        return strConversion;
      }
      else
      {
        return m.Value;
      }

    }


    protected abstract string replacedTokenValue(string strObjectName, string strPropertyName, string strFormat);

  }
}